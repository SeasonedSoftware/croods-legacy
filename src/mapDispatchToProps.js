import { bindActionCreators } from 'redux'

import createActions from './createActions'

export default (dispatch, { options, ...props }) => ({
  actions: bindActionCreators(
    createActions({ ...options, ...props }),
    dispatch,
  ),
})
