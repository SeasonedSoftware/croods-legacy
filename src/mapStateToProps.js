import find from 'lodash/find'
import get from 'lodash/get'

export default (state, { name, parentId }) => {
  const statePart = get(state, name)
  const currentState = find(statePart, block => block.parentId === parentId)
  return get(currentState, 'state') || {}
}
