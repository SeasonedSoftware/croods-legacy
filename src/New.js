import React, { Component, Fragment } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import customPropTypes from './customPropTypes'
import providerProps from './providerProps'
import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import renderIfPresent from './renderIfPresent'

class New extends Component {
  constructor(props) {
    super(props)
    const { createError, actions, renderCreated } = props
    if (renderCreated) {
      console.warn(
        'renderCreated is deprecated and will be removed in future versions of Croods. Please, update your code to use afterCreate function instead. [Docs at: https://croods-docz.netlify.com/new]',
      )
    }

    createError && actions.resetCreateError()
  }

  componentDidUpdate(prevProps) {
    const { actions, created } = this.props
    const { created: oldCreated } = prevProps

    if (created && !oldCreated) {
      actions.resetCreated()
    }
  }

  render() {
    const { render, renderCreated, actions, created, creating } = this.props
    const { createError: error } = this.props
    const { create } = actions

    return (
      <Fragment>
        {renderIfPresent(renderCreated, created)}
        {render({ create, creating, error }, this.props)}
      </Fragment>
    )
  }
}

New.propTypes = {
  /** Defines the deep properties used in the component. Ex: foo.bar */
  name: customPropTypes.name.isRequired,
  /** A function returning a React Node. Ex: ({ create, creating = bool, error }, props) -> < JSX /> */
  render: PropTypes.func.isRequired,
  /** A function returning a React Node. Ex: (created = {}) -> < JSX /> */
  renderCreated: PropTypes.func,
  /**  Parent element's Id, for more specificity in changes.  */
  parentId: customPropTypes.id,
  /**  Ex: /foo/bar-_12?&= */
  path: customPropTypes.path,

  ...providerProps,
}

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(New),
)
