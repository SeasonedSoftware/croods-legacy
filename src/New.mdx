---
name: New
---

import { Playground, PropsTable } from 'docz'
import { NewDocs } from './New'

# New

Thats the new component

<PropsTable of={NewDocs} />