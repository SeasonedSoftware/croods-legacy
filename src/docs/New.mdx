---
name: New
route: /new
---

import { Playground } from 'docz'
import New from '../New'

# New

The New component provides a way to create a resource and show or use the response data.


## API

| Property      | Type               | Required | Default    | Description                                                                                          |
| ------------- | ------------------ | -------- | ---------- | ---------------------------------------------------------------------------------------------------- |
| name          | String             | true     | -          | It must be equal to the reducer name. Defines the deep properties used in the component. Ex: foo.bar |
| render        | Func               | true     | -          | A function returning a React Node. Ex: ({ create, creating = bool, error }, props) -> < JSX />       |
| renderCreated | Func               | false    | -          | A function returning a React Node. Ex: (created = {}) -> < JSX />                                    |
| parentId      | String/Number      | false    | -          | Parent element's Id, for more specificity in changes.                                                |
| path          | String             | false    | `/${name}` | Ex: /foo/bar-\_12?&=                                                                                 |
| method        | Enum (post or put) | false    | `post`     | The http request method.                                                                             |

## How to use

You have to add a reducer with the same name that you will pass as prop to the New component.

```javascript
import { createReducer } from 'croods'

const reducers = {
  foobar: createReducer('foobar'),
}
```

### render

The render function receives three important parameters, **create** (func), **creating** (bool) and **error** (string).

- The **create** function makes a `POST` request on `path` with the JSONified object parameter as body.
- The **creating** is a boolean that will be `true` while requesting and will switch back to `false` once the request succeeds or fails.
- The `error` is populated after calling `create`, only if the server response has an error status code.

### renderCreated

The renderCreated function is fired when the `create` function is called and the server response is of 2xx status.
It receives one parameter: `created` containing the server's response body.

## Example

You can see a working sample of this component's usage [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/New.js).
