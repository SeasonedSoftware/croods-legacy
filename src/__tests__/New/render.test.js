import React from 'react'
import { connect } from 'react-redux'
import renderer from 'react-test-renderer'

import New from '../../New'
import mapStateToProps from '../../mapStateToProps'
import mapDispatchToProps from '../../mapDispatchToProps'

jest.mock('../../withOptions', () => Component => Component)

jest.mock('react-redux', () => ({
  connect: jest.fn(() => Component => Component),
}))

const render = jest.fn((newProps, props) => (
  <div {...newProps} {...props}>
    New
  </div>
))

const actions = { resetCreated: jest.fn() }

beforeEach(() => {
  render.mockClear()
  actions.resetCreated.mockClear()
})

it('renders correctly', () => {
  const props = { render, actions }
  const tree = renderer.create(<New {...props} />).toJSON()

  expect(tree).toMatchSnapshot()
  expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
  expect(actions.resetCreated).not.toHaveBeenCalled()
})

describe('with renderCreated and without created', () => {
  it('renders edit', () => {
    const renderCreated = jest.fn(created => <div>Created: {created}</div>)
    const props = { render, actions, renderCreated }
    const tree = renderer.create(<New {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('with renderCreated and created', () => {
  it('renders created', () => {
    const renderCreated = jest.fn(created => <div {...created}>Created </div>)
    const props = { render, actions, renderCreated, created: { foo: 'bar' } }
    const tree = renderer.create(<New {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).not.toHaveBeenCalled()
  })
})

describe('without renderCreated and with created', () => {
  it('renders new', () => {
    const props = { render, actions, created: { foo: 'bar' } }
    const tree = renderer.create(<New {...props} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
