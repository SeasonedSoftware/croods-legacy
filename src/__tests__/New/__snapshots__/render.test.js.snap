// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  actions={
    Object {
      "resetCreated": [Function],
    }
  }
  create={undefined}
  creating={undefined}
  error={undefined}
  render={[Function]}
>
  New
</div>
`;

exports[`with renderCreated and created renders created 1`] = `
<div
  foo="bar"
>
  Created 
</div>
`;

exports[`with renderCreated and without created renders edit 1`] = `
<div
  actions={
    Object {
      "resetCreated": [Function],
    }
  }
  create={undefined}
  creating={undefined}
  error={undefined}
  render={[Function]}
  renderCreated={[Function]}
>
  New
</div>
`;

exports[`without renderCreated and with created renders new 1`] = `
<div
  actions={
    Object {
      "resetCreated": [Function],
    }
  }
  create={undefined}
  created={
    Object {
      "foo": "bar",
    }
  }
  creating={undefined}
  error={undefined}
  render={[Function]}
>
  New
</div>
`;
